tests/common/util.rs

```diff
@@ -1189,14 +1239,284 @@ impl UCommand {
+        let mut buffer = Vec::<u8>::new();
+        let mut stdout = self.output.reopen().unwrap();
+
+        stdout.read_to_end(&mut buffer).unwrap();
+        buffer
+    }
+
+    pub fn output_exact_bytes(&self, size: usize) -> Vec<u8> {
+        let mut buffer = vec![0; size];
+        let mut output = self.output.reopen().unwrap();
+
+        output.read_exact(&mut buffer).unwrap();
+        buffer
+    }
+}
+
+pub struct UChild {
+    raw: Child,
+    bin_path: String,
+    util_name: Option<String>,
+    tmpd: Option<Rc<TempDir>>,
+    output: Option<CombinedOutput>,
+}
+
+impl UChild {
+    fn new(
+        child: Child,
+        bin_path: String,
+        util_name: Option<String>,
+        tmpd: Option<Rc<TempDir>>,
+        output: Option<CombinedOutput>,
+    ) -> Self {
+        Self {
+            raw: child,
+            bin_path,
+            util_name,
+            tmpd,
+            output,
+        }
+    }
+
+    /// Returns a mutable reference of the wrapped child process.
+    pub fn as_child(&mut self) -> &mut Child {
+        &mut self.raw
+    }
+
+    /// TODO: Sends SIGKILL (9) to the child process and terminates the child unconditionally. If
+    /// `ignore_errors` is true unwrap() can be safely used.
+    ///
+    /// # Error If `ignore_errors` is false then the error from [`Child::kill`] is returned.
+    pub fn kill(&mut self, ignore_errors: bool) -> io::Result<()> {
+        match self.raw.kill() {
+            Ok(_) => Ok(()),
+            Err(_) if ignore_errors => Ok(()),
+            Err(error) => Err(error),
+        }
+    }
+
+    // TODO: use wait_with_output.unwrap() and return plain CmdResult instead?
+    /// Wait for the child process to terminate and return a [`CmdResult`]. This method can also be
+    /// run if the child process was killed with [`UChild::kill`].
+    ///
+    /// # Errors
+    /// If [`Child::wait_with_output`] returns an error this error will be returned
+    pub fn wait(self) -> io::Result<CmdResult> {
+        let command_output = self.raw.wait_with_output()?;
+
+        let (stdout, stderr) = match self.output {
+            Some(output) => (output.output_as_bytes(), vec![]),
+            None => (command_output.stdout, command_output.stderr),
+        };
+
+        Ok(CmdResult {
+            bin_path: self.bin_path.clone(),
+            util_name: self.util_name.clone(),
+            tmpd: self.tmpd.clone(),
+            code: command_output.status.code(),
+            success: command_output.status.success(),
+            stdout,
+            stderr,
+        })
+    }
+
+    pub fn stdout(&mut self) -> String {
+        String::from_utf8(self.stdout_bytes()).unwrap()
+    }
+
+    /// Reads and returns all bytes from child's stdout. If `stderr` is not redirected to `stdout`
+    /// with [`UCommand::stderr_to_stdout`] and the child process is still running then this
+    /// method will block indefinitely.
+    pub fn stdout_bytes(&mut self) -> Vec<u8> {
```

Comment:
would love to do that, but all this blocking of stdout and stderr is inherent to the way we capture output from the child via pipe to the master process. Currently, only the `..exact..` methods can be used to capture the output without blocking as long as the child is still running. As a sidenote: If we would use a file per default in set_stdout and set_stderr like we do with stderr_to_stdout all these blocking methods would not block anymore.

Reply:
What if a method like `stdout_is` would implicitly do a read up to the length of its input to a buffer and check that?

Reply:
Wait, actually I don't quite understand. Why can't we use `ChildStdout::read` instead of `ChildStdout::read_exact` to make it non-blocking if we don't want an exact byte count? I suppose there is a possibility to not get the entire thing or something?

Reply:
> Would there be any downsides to using a file like you suggest?

I indeed think this is the far better option. I'm not a fan of these exact reads, at all. I've used it for now because this was the established method to read from the pipe, if you look at the methods and functions which are already there. I also noticed that the read_exact method isn't very close to the real world, which we actually test against. 

Let's take for example the test `test_tail::test_follow_single`. If you try it on the command line `target/debug/tail -f file_with_data_but_no_newlines` then there is no output but there definitely should be the output of the file visible. So the test is wrong in passing the assertions. I've fixed this error in #3952. The fix is to flush stdout or a writer with stdout after io::copy etc. Actually, the much bigger problem is, that we don't notice such things in the tests because of this exact reading method in which read is called continuously until the buffer is filled. So, yeah I would prefer using files as default for stderr and stdout instead of Stdio::piped() in run_no_wait.

Reply:
... and remove these exact reading methods from UChild

Reply:
>  Why can't we use ChildStdout::read instead of ChildStdout::read_to_end to make it non-blocking? I suppose there is a possibility to not get the entire output or something?

Yes, as far as I've understood `read`, `read` doesn't make too much promises about the amount of bytes read from the source. And `read` also needs a fixed size buffer. However, I'll try both, the `read` method and the file method in run_no_wait.

Reply:
I gave the change from the default `Stdio::piped()` to file based handling in `run_no_wait` a go. I haven't refactored most of the tests yet, so the ci is failing. I ran all tests on a linux platfrom and there were around 60 failing tests. All of them are using the old `run_no_wait` method without `UChild` and are expected to fail right now. Note, that the usual `run` method still works like before, so we are backwards compatible. I also changed some additional behavior:

* each call to `assert_now` or the `stdout()` and `stderr()` methods are consuming the output, so each subsequent call to any of these methods returns the subsequent output. This matches the old behavior were the output was piped and should be the most intuitive default behavior.
* the new consuming behavior is also the default when using `stderr_to_stdout`, so there's no change in the consuming behavior anymore, using this method or not.

What's also cool about using files is that we have access to the complete output if we wish to do so and nothing is really lost. I just haven't implemented it, yet. What's your opinion?

Reply:
I like it! Though I think we have to make some simplification steps from here. That is, evaluate the functionality and methods that are actually useful for writing tests.